### PR TITLE
separate transport-specific streaming logic out

### PR DESCRIPTION
**Pull request by phryneas**
See a lot of comments further down :)

### CODE REVIEW

**Comment by phryneas** [comment]
*sigh*

**Review by phryneas — COMMENTED** [review]
This should be in a good state to be reviewed.

In a second PR, I'm gonna split this up into multiple actual packages, but this is more than enough for one PR.

File: package/src/combined.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export * from "./rsc/index.js";
+export * from "./ssr/index.js";
+
+export {};
```
[the remark this answers is not in the document]
Comment (jerelmiller) [reply package/src/combined.ts:4]: Would it be worth adding this as a code comment? I know I'm going to forget this in 6 months 😂

**Comment by luchillo17** [comment]
I've started checking on these changes and there's a lot to look at, but I see the description of the PR is very small, it would be nice to know what is the reason this PR exists, the title doesn't tell me much, is it a refactor for DRY principles? or is there a functional change in how data and/or UI are streamed to the client side?

**Comment by phryneas** [comment]
@luchillo17 We are in the process of separating this package into two separate ones - one more generic `@apollo/client-react-streaming` and the more specific `@apollo/experimental-nextjs-app-support`. This is necessary because other frameworks (e.g. RedwoodJs) are now adapting streaming SSR, and we want to add easier support for those.

This is the first of multiple PRs, and the follow-up PR with the actual package split is going over in https://github.com/apollographql/apollo-client-nextjs/pull/193

File: package/src/combined.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (phryneas) [reply package/src/combined.ts:4]: Good call! I added a comment over in #193

File: package/src/ssr/WrappedApolloClient.test.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    useStaticValueRefStub = () => ({ current: FIRST_HOOK_RESULT as any });
+    simulateRequestData!(FIRST_WRITE);
+
+    await new Promise((resolve) => setTimeout(resolve, 1000));
+
+    await findByText("User");
+
+    assert.ok(attemptedRenderCount > 0);
+    assert.ok(finishedRenderCount > 0);
+  }
```
Comment (jerelmiller) [inline package/src/ssr/WrappedApolloClient.test.tsx:200]: Sorry if this is a naive question, but does the cache play any role here? As in, do we need to check the cache to make sure the result made it in the cache?

**Review by jerelmiller — APPROVED** [review]
I'm not going to pretend I understand all of this 100% but I've learned a lot looking through this. I like a lot of the refactoring that has been done here and think this is a solid base moving forward. I've held this up long enough though, so I'll continue to dig deeper on my own! Incredible work 🎉

File: package/src/rsc/registerApolloClient.tsx [hunk 3]
```diff
@@ -1,10 +1,10 @@
 import "server-only";
 
-import type { ApolloClient } from "@apollo/client";
-import * as React from "react";
+import type { ApolloClient } from "@apollo/client/index.js";
+import { cache } from "react";
```
[the remark this answers is not in the document]
Comment (phryneas) [reply package/src/rsc/registerApolloClient.tsx:4]: Yeah, but I also drop out rehackt as a dependency for this in the next PR :)

File: package/src/ssr/Manual/ManualDataTransport.tsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      <DataTransportContext.Provider
+        value={useMemo(
+          () => ({
+            useStaticValueRef,
+          }),
+          [useStaticValueRef]
+        )}
+      >
+        {children}
+      </DataTransportContext.Provider>
```
[the remark this answers is not in the document]
Comment (phryneas) [reply package/src/ssr/Manual/ManualDataTransport.tsx]: True!
This had a lot more functions at some point :)

File: package/src/ssr/WrappedApolloClient.test.tsx (hunk 2 above)
Comment (phryneas) [reply package/src/ssr/WrappedApolloClient.test.tsx:209]: Doesn't hurt - I've added another assertion :)

**Comment by dac09** [comment]
🥳🥳🥳🥳🥳

**Pull request merged**